server/storage/backend/backend.go

```diff
@@ -499,6 +499,7 @@ func (b *backend) defrag() error {
 	tdbp := temp.Name()
 	tmpdb, err := bolt.Open(tdbp, 0600, &options)
 	if err != nil {
+		temp.Close()
```

Comment:
we also need to remove the temp file, refer to below https://github.com/etcd-io/etcd/blob/4ad92611942a842c06324ae2da74ead48db79ee9/server/storage/backend/backend.go#L526

Reply:
Is this really needed? etcd anyway removes the `db.tmp.*` files at startup.

Reply:
It's the tmp file instead of db.tmp.*

Reply:
So should I simply add a `os.Remove(temp.Name())` after the call to `Close`?

Reply:
Yes, I think so. Please also log an error message if the removing fails

Reply:
I added something, let me know what you think. Is there also a need to write a unit test for this?

In my opinion the error log doesn't really provide any value to an etcd user so personally I would omit it. As a user I do not really care if there is an old temp file lying around (that will be deleted on next etcd restart anyway) that is why I think it is not needed. However, I'm fine with adding it if you think it is needed 🙂

Reply:
Is it possible to use defer here for cleanup? I saw the line 525 does the same thing. If there are multiple places to handle, maybe we can consider to use defer here.

Reply:
> Is it possible to use defer here for cleanup? 

We don't need to close & remove it in normal case. The tmp will be renamed to the bbolt db file.




> In my opinion the error log doesn't really provide any value to an etcd user so personally I would omit it. 

It's for debug purpose. It should be very rare. Overall not a big problem, and doesn't deserve too much time to debate on this.

Reply:
> We don't need to close & remove it in normal case. The tmp will be renamed to the bbolt db file.

I mean, 

```
defer func() {
    if returnErr != nil {
          dosomething
    }
}()
```

It's just an option.